proposals/automating-releases.md

```diff
@@ -0,0 +1,185 @@
+1. This proposal encompasses all binaries/images currently built for an operator-sdk release.
+Once Ansible- and Helm-related code are split into their respective repos, will
+this release process follow?
+1. Continue using our [changelog generator][sdk-changelog-generator], or use goreleaser's
+[generator][goreleaser-changelog]?
+
+## Summary
+
+This proposal outlines an automated release pipeline for the operator-sdk repository
+that includes both binary and image builds. The [`goreleaser`][goreleaser] release tool
+will replace the existing combination of the release script `release.sh`, manual `git`
+steps, and manual release publishing with a one "click" solution. To users, no changes
+will occur since the release artifacts are not changing. To developers, the release
+process will become much simpler, hopefully to the point where a new contributor can
+run the release with little process knowledge.
+
+## Motivation
+
+Releasing the operator-sdk repo currently requires several steps that are mostly manual,
+with some automation around building binaries for all architectures supported. Tools like
+[`goreleaser`][goreleaser] can automate this process entirely by tagging code, and building
+and publishing binaries and images with one "click".
+
+## Goals
+
+- Remove all manual steps from the release process
+- Centralize release configuration
+- One-click releasing
+
+### Non-Goals
+
+- Change the content or constituency of output release artifacts in any way
+- Change the testing process in any way
+- Change our CI provider
+
+## Proposal
+
+### User Stories
+
+#### Story 1 - Release the operator-sdk repo in one click
+
+As a contributor I want to run the release in as few operations as possible.
+This feature will enable new contributors to release the repo easily, instead
+of relying on a long onboarding document and the expertise of a few contributors.
+
+#### Story 2 - Easily modify release configuration
+
+As a contributor I want to easily improve the release configuration whenever necessary,
+which includes adding new release architectures. Configuration should be documented,
+comprehensible, and centralized.
+
+### Implementation Details/Notes/Constraints
+
+The [`goreleaser`][goreleaser] release tool satisfies the above goals and user stories:
+- Manages code tags
+- Builds and pushes [binaries][goreleaser-build] and [images][goreleaser-docker]
+  - Applies to `operator-sdk`, `ansible-operator`, and `helm-operator`
+- Centralized configuration and a well-documented [configuration spec][goreleaser-config]
+- Handles the full release process with one click
+
+`goreleaser` configuration is CI-agnostic, so if we choose to switch CI systems in the future
+we can bring our configuration with us verbatim.
+
+Additionally it supports generating release notes based on commits, or from a
+[custom source][goreleaser-custom-changelog]; we currently use a custom changelog generator
+which we can continue using.
```

Comment:
If we go to commit based changelog, we will need a much stricter review process for commit messages. The only thing I saw that stood out is that ours has the ability to flag a breaking change and help generate the migration guide.

Reply:
I think a migration guide generator is a must-have.

Reply:
How would commit messages used generating docs interact with the squash-and-merge button?

Reply:
IMO, if we go with commit based changelog, we would want to enforce a format via a git hook or a check in CI or something.

One thing that is problematic about our current process is that the CHANGELOG and migration guide generation happens as a pre-step to the tag, which requires a PR and a CI run. Ideally the solution we come up with would remove the pre-release PR entirely, such that we can literally just tag `master` at any time to make the release.